server/src/main/java/org/elasticsearch/index/mapper/SyntheticSourceDocumentParserListener.java

```diff
@@ -0,0 +1,192 @@
+            this.parentMapper = parentMapper;
+
+            // TODO use actual value from initial parser (add a "document start" token with metadata?)
+            this.data = XContentBuilder.builder(XContentType.JSON.xContent());
+
+            this.depth = 0;
+
+            consume(startingToken);
+        }
+
+        public State consume(Token token) throws IOException {
+            switch (token) {
+                case Token.StartObject ignored -> {
+                    data.startObject();
+                    if (startingToken instanceof Token.StartObject) {
+                        depth += 1;
+                    }
+                }
+                case Token.EndObject ignored -> {
+                    data.endObject();
+
+                    if (startingToken instanceof Token.StartObject) {
+                        depth -= 1;
+                        if (depth == 0) {
+                            var parentOffset = parentMapper.fullPath().length() + 1;
+                            // TODO the way we store values is not final, maybe we should put them directly in DocumentParserContext ?
+                            // does not feel great though
+                            valuesToStore.add(
+                                new IgnoredSourceFieldMapper.NameValue(fullPath, parentOffset, XContentDataHelper.encodeXContentBuilder(data), document)
+                            );
+                            return returnState;
+                        }
+
+                    }
+                }
+                case Token.StartArray ignored -> {
+                    data.startArray();
+                    if (startingToken instanceof Token.StartArray) {
+                        depth += 1;
+                    }
+                }
+                case Token.EndArray ignored -> {
+                    data.endArray();
+
+                    if (startingToken instanceof Token.StartArray) {
+                        // TODO extract function
+                        depth -= 1;
+                        if (depth == 0) {
+                            var parentOffset = parentMapper.fullPath().length() + 1;
+                            // does not feel great though
+                            valuesToStore.add(
+                                new IgnoredSourceFieldMapper.NameValue(fullPath, parentOffset, XContentDataHelper.encodeXContentBuilder(data), document)
+                            );
+                            return returnState;
+                        }
+                    }
+                }
+                case Token.FieldName fieldName -> data.field(fieldName.name());
+                case Token.StringValue stringValue -> data.value(stringValue.value());
+                case Token.BooleanValue booleanValue -> data.value(booleanValue.value());
+                case Token.IntValue intValue -> data.value(intValue.value());
+            }
+
+            return this;
+        }
+    }
+
+    class Watching implements State {
```

Comment:
As usual, some comments would help here.. I assume this is the one iterating over the doc, tracking its structure and identifying cases that need to be stored.